### PR TITLE
Type hint multiple accumulator variables

### PR DESCRIPTION
**Pull request by adamtheturtle**
This is required for strict type checking, and even without that, it prevents errors of e.g. `.append`-ing the wrong type accidentally.

### CODE REVIEW

**Comment by chrisjsewell** [comment]
> This is required for strict type checking

Well.. technically I don’t believe it is required; mypy already computes the list type, based on the first append, so you only need to do this if you have covariant types in your list (which must will already complain about without the typing)

bit to say I’m against it, but it does feel a little bit, adding spurious code for the sake of it 😬

**Comment by adamtheturtle** [comment]
I should have said, I'm using `pyright` as well as `mypy` in my dev environment and trying to get towards 0 errors. I appreciate that there is no agreement to use `pyright` on this codebase, nor am I pushing for it.

I choose to use `pyright` as well partly because that's what VSCode's Pylons extension uses, and I have that installed, so `pyright`-satisfying changes give me better highlighting and autocomplete.

I won't be offended if this is closed!

File: sphinx/domains/cpp/_parser.py [hunk 1]
```diff
@@ -1788,7 +1788,7 @@ def _parse_template_parameter(self) -> ASTTemplateParam:
             except DefinitionError as eNonType:
                 self.pos = pos
                 header = "Error when parsing template parameter."
-                errs = []
+                errs: list[tuple[DefinitionError, str]] = []
                 errs.append(
```
Comment (picnixz) [inline sphinx/domains/cpp/_parser.py:1792]: Here, it'd be better to just write something like `errs = [...]` without those un-necessary append calls.

File: sphinx/domains/cpp/_parser.py [hunk 2]
```diff
@@ -2090,7 +2090,7 @@ def parse_xref_object(self) -> tuple[ASTNamespace | ASTDeclaration, bool]:
                 self.assert_end()
                 return res2, False
             except DefinitionError as e2:
-                errs = []
+                errs: list[tuple[DefinitionError, str]] = []
```
Comment (picnixz) [inline sphinx/domains/cpp/_parser.py:2093]: ditto

File: sphinx/domains/cpp/_parser.py [hunk 3]
```diff
@@ -2112,7 +2112,7 @@ def parse_expression(self) -> ASTExpression | ASTType:
                 return typ
             except DefinitionError as exType:
                 header = "Error when parsing (type) expression."
-                errs = []
+                errs: list[tuple[DefinitionError, str]] = []
```
Comment (picnixz) [inline sphinx/domains/cpp/_parser.py:2115]: ditto

File: sphinx/writers/html5.py [hunk 4]
```diff
@@ -872,7 +872,9 @@ def visit_table(self, node: Element) -> None:
         self._table_row_indices.append(0)
 
         atts = {}
-        classes = [cls.strip(' \t\n') for cls in self.settings.table_style.split(',')]
+        classes: list[str] = [
```
Comment (picnixz) [inline sphinx/writers/html5.py:875]: Is it required because `settings.table_sync` has type Any? Does `reveal_type(classes)` for mypy is `list[Any]`?

**Review by picnixz — COMMENTED** [review]
I'm on the same opinion as Chris here and think it's a pyright issue (or feature, call it however you want).

There are however places where the append() calls are not necessary and should be replaced by a list literal construction.

**Comment by adamtheturtle** [comment]
Thank you @picnixz , I'm happy to make changes, if you feel they are valuable, and would be merged.

**Comment by picnixz** [comment]
- I think we can close this specific PR.
- But you can open a new one to remove redundant `append()`.

**Comment by adamtheturtle** [comment]
Thank you. Perhaps one day this gets re-opened if Sphinx adopts pyright.

**Comment by AA-Turner** [comment]
@adamtheturtle how feasible is adopting pyright now? I think it's a goal we should aim towards -- would you be willing to write up a tracking issue/PR?

A

**Comment by adamtheturtle** [comment]
> how feasible is adopting pyright now?

Honestly I'm not sure, and it depends on where the goalposts are - `strict` mode? `ignore`s welcome to what extent? etc. etc.

As the maintainer of a Sphinx extension, my goal is to have `pyright --verifytypes` work on my project, which means that *a lot* would have to be typed.
In my first attempt, `pyright` showed ~3.5k issues IIRC and I think that this PR reduced that by ~10%.

> would you be willing to write up a tracking issue/PR?

For sure not a single PR to have `pyright` completely happy.
Yes, I'm happy to have a go at setting `pyright` up in CI, and creating an issue, with many files ignored with a `pyright` config to exclude many files, and an issue with checkboxes for removing each of those files.
I'm not sure how much I could commit to ticking those off - partly time/effort but mostly due to my level of understanding.

If you think that would be valuable, I'm happy to do it. My guess is that with folks including @danieleades toiling away at getting `mypy --strict` working, the best way forward is to do that first and then `pyright` will be an easier job.

**Comment by danieleades** [comment]
@adamtheturtle mypy was introduced (is being introduced) incrementally. Could a similar approach be used with pyright?

**Comment by adamtheturtle** [comment]
Yes, with the incremental approach I was thinking of being:

* Put `pyright` in CI
* Exclude most files (or maybe most rules?)
* Create an issue with each excluded file/rule as a checkbox
* One by one, remove the exclude

How is the `mypy` introduction incremental? In a similar way?

**Comment by danieleades** [comment]
> Yes, with the incremental approach I was thinking of being:
> 
> * Put `pyright` in CI
> * Exclude most files (or maybe most rules?)
> * Create an issue with each excluded file/rule as a checkbox
> * One by one, remove the exclude
> 
> How is the `mypy` introduction incremental? In a similar way?

Yeah more or less the same. Incrementally add rules, incrementally shrink the 'whitelist' of ignored files.

The majority of the work has actually been upstreaming better type stubs into typeshed for docutils (also very much a work in a progress)

**Pull request closed without merging**